src/services/document-service.js

```diff
@@ -342,23 +262,133 @@ class DocumentService {
+        }
+      } else {
+        existingDocumentRevisions = [];
+        ancestorRevision = null;
+      }
+
+      const nextOrder = await this.documentOrderStore.getNextOrder();
+      const newSections = mapNewSections({ sections: doc.sections, ancestorSections: ancestorRevision?.sections, restoredFrom });
+      const newDocumentRevision = this._buildDocumentRevision({ doc, revisionId, documentKey, userId, nextOrder, restoredFrom, sections: newSections });
+      const saveOptions = databaseSession ? { session: databaseSession } : {};
+
+      logger.info(`Saving new document revision with id ${newDocumentRevision._id}`);
+      await this.documentRevisionStore.save(newDocumentRevision, saveOptions);
+
+      const latestDocument = this._createDocumentFromRevisions([...existingDocumentRevisions, newDocumentRevision]);
+
+      logger.info(`Saving latest document with revision ${latestDocument.revision}`);
+      await this.documentStore.save(latestDocument, saveOptions);
+
+      return newDocumentRevision;
+
+    } finally {
+      if (lock) {
+        await this.documentLockStore.releaseLock(lock);
+      }
+    }
+  }
+
+  _createNewSections({ sections, ancestorSections = [], restoredFrom = null }) {
+    return sections.map(section => {
+      const sectionKey = section.key;
+      const ancestorSection = ancestorSections.find(s => s.key === sectionKey) || null;
+
+      if (ancestorSection) {
+        logger.info(`Found ancestor section with key ${sectionKey}`);
+
+        if (ancestorSection.type !== section.type) {
+          throw new Error(`Ancestor section has type ${ancestorSection.type} and cannot be changed to ${section.type}`);
+        }
+
+        if (ancestorSection.deletedOn && section.content) {
+          throw new Error(`Ancestor section with key ${sectionKey} is deleted and cannot be changed`);
+        }
+
+        // If not changed, re-use existing revision:
+        if (deepEqual(ancestorSection.content, section.content)) {
+          logger.info(`Section has not changed compared to ancestor section with revision ${ancestorSection.revision}, using the existing`);
+          return cloneDeep(ancestorSection);
+        }
+      }
+
+      if (!section.content && !restoredFrom) {
+        throw new Error('Sections that are not deleted must specify a content');
+      }
+
+      logger.info(`Creating new revision for section key ${sectionKey}`);
+
+      // Create a new section revision:
+      return {
+        revision: uniqueId.create(),
+        key: sectionKey,
+        deletedOn: null,
+        deletedBy: null,
+        deletedBecause: null,
+        type: section.type,
+        content: section.content && cloneDeep(section.content)
+      };
+    });
+  }
+
+  _copySections({ sections }) {
+    return sections.map(section => ({
+      revision: section.revision,
+      key: section.key,
+      deletedOn: null,
+      deletedBy: null,
+      deletedBecause: null,
```

Comment:
What happens in case we import a document with sections that have been deleted?

Reply:
D'oh, we loose this info, but of course.
Will change to do a `sections.map(section => { ...section })` then.
I'd still keep this method though so that `_createDocumentRevision` doesn't need to do in internal `if`-`else`

Reply:
or just `return sections.map(s => cloneDeep(s));`

Reply:
I agree on the no `if`-`else`

Reply:
Changed!